asdf/_block/writer.py

```diff
@@ -0,0 +1,105 @@
+        self.compression_kwargs = compression_kwargs
+
+    @property
+    def data(self):
+        if callable(self._data):
+            return self._data()
+        return self._data
+
+    @property
+    def data_bytes(self):
+        data = self.data
+        if data is not None:
+            return np.ndarray(-1, np.uint8, data.ravel(order="K").data)
+        return np.ndarray(0, np.uint8)
+
+
+def write_blocks(fd, blocks, padding=False, streamed_block=None, write_index=True):
+    """
+    Write a list of WriteBlocks to a file
+
+    Parameters
+    ----------
+    fd : file or generic_io.GenericIO
+        File to write to. Writing will start at the current position.
+
+    blocks : list of WriteBlock
+        List of WriteBlock instances used to get the data and options
+        to write to each ASDF block.
+
+    padding : bool or float, optional, default False
+        If False, add no padding bytes between blocks. If True
+        add some default amount of padding. If a float, add
+        a number of padding bytes based off a ratio of the data
+        size.
+        See ``asdf._block.io.write_block`` ``padding`` argument for
+        more details.
+
+    streamed_block : WriteBlock, optional
+        If provided (not None) include this WriteBlock as
+        the final block in the file and mark it as a streamed
+        block.
+
+    write_index : bool, optional, default True
+        If True, include a block index at the end of the file.
+        If a streamed_block is provided (or the file is not
+        seekable) no block index will be written.
+
+    Returns
+    -------
+    offsets : list of int
+        Byte offsets (from the start of the file) where each
+        block was written (this is the start of the block magic
+        bytes for each block). This list includes the offset of
+        the streamed_block if it was provided.
+        If the file written to is not seekable these offsets
+        will all be None.
+
+    headers : list of dict
+        Headers written for each block (including the streamed_block
+        if it was provided).
+    """
+    offsets = []
+    headers = []
+    for blk in blocks:
+        if fd.seekable():
+            offset = fd.tell()
+        else:
+            offset = None
+        offsets.append(offset)
+        fd.write(constants.BLOCK_MAGIC)
+        headers.append(
+            bio.write_block(
+                fd,
+                blk.data_bytes,
+                compression_kwargs=blk.compression_kwargs,
+                padding=padding,
+                compression=blk.compression,
+            )
+        )
+    if streamed_block is not None:
+        if fd.seekable():
+            offset = fd.tell()
+        else:
+            offset = None
+        offsets.append(offset)
+        fd.write(constants.BLOCK_MAGIC)
+        headers.append(bio.write_block(fd, streamed_block.data_bytes, stream=True))
+    elif len(offsets) and write_index and fd.seekable():
```

Comment:
Exactly!

While attempting to write a non-seekable file test I ran into this:
https://github.com/asdf-format/asdf/issues/1542
and:
https://github.com/asdf-format/asdf/issues/1545

I thought this could be mostly handled by not calling tell on a non-seekable file (which meant that no block index could be generated for a non-seekable file). However it appears more complicated as there are also files that are non-seekable where tell returns a valid result:
https://github.com/asdf-format/asdf/issues/1552

I will work on addressing hopefully all 3 of these in this PR. My current thoughts are:
- modify generic_io to catch the `Illegal seek` returned from tell and instead return None
- modify the block index generation to use tell for non-seekable files and to not write out an index if it contains all 0s (because of the windows bug which has wider implications) or all Nones

Reply:
I believe I addressed the non-seekable file write issues here:
684ccf375ab6dfb99b61a625d2a53b7f30a62082
the above commit also adds tests for the various odd `tell` results.